### ISSUE TITLE
The new NonNullable type doesn't work the same when combined with index types

### ISSUE THREAD

**Issue by niieani** [microsoft/TypeScript#50462]
# Bug Report

### 🔎 Search Terms

NonNullable index

### 🕗 Version & Regression Information

- This changed between versions 3.7.4 and 3.8.1-rc

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/C4TwDgpgBAogJgcwgHgCrggOwIYFtoQAewWcAzlGcAE4CWmCUAvJTfQgHzNQDeAUFCjUI2OAHtMAGxBQIANyzAyALigAlEeKkgAgtWrYQyHkM0TpUAPqXQkHPlXo7eaAF8oAHyiYArpMkcANx8rsF8RGBi1MBQttAAkpiYENQwCpjATlguyPBIqLLEpBR5EFwsAHISFX6S2ABGkiilqADaAOTyimTtALqtvrj1Kb0cHdZx9hB9wUA)

### 💻 Code

```ts
type Edge<Typename extends string = string> = {
  readonly events: ReadonlyArray<{ readonly __typename: Typename } | null>;
};

export type InnerEventTypename<EdgeT extends Edge> = NonNullable<EdgeT['events'][number]>['__typename'];

```

### 🙁 Actual behavior

Error: Type '"__typename"' cannot be used to index type 'NonNullable<EdgeT["events"][number]>'.

### 🙂 Expected behavior

No error.

**Comment by RyanCavanaugh** [comment]
Simplified
```ts
type A = {
    prop: { x: string, y: string } | null
};
// 4.7: OK
// 4.8: Error
type Test<T extends A> = NonNullable<T['prop']>['x'];
```

**Comment by ahejlsberg** [comment]
This was an issue only in `4.8.0-beta`. It was fixed in `4.8.1` and doesn't currently repro.

**Comment by RyanCavanaugh** [comment]
I oversimplified, then. This is (newly) erroring in 4.8.2
```ts
type Column<T = unknown> = (keyof T extends never ? { id?: number | string } : { id: T }) & {
    title?: string;
}

// no error if `key: keyof Column`
function getColumnProperty<T>(column: Column<T>, key: keyof Column<T>) {
    return column[key];
}
```

**Comment by RyanCavanaugh** [comment]
Actually since OP *is* fixed, let's track over at that one - #50465
